File: 000001-000500/000015.三数之和/15.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
// 方法一：用时 120ms
// 执行用时超过了 100%
// 消耗内存超过了 71%
var threeSum = function(nums) {
  if (nums.length < 3) { return [] }
  nums = nums.sort((a, b) => (a - b))
  let result = []
  for (let i = 0; i < nums.length - 2 && nums[i] <= 0; i++) {
    if (nums[i] > 0) { break }
    if (i > 0 && nums[i] === nums[i - 1]) { continue }
    let subsum = 0 - nums[i]
    let left = i + 1
    let right = nums.length - 1
    // console.log(`i=${i} nums[i]=${nums[i]} subsum=${subsum}`)
    while (left < right) {
      // console.log(`left=${left} right=${right} nums[left+right]=${nums[left]+nums[right]} subsum=${subsum} `, nums[left]+nums[right]==subsum)
      if (nums[left] + nums[right] == subsum) {
        result.push([nums[i], nums[left], nums[right]])
        left ++
        right --
        while (nums[left] === nums[left - 1]) {
          left ++
        }
        while (nums[right] === nums[right + 1]) {
          right --
        }
      } else if (nums[left] + nums[right] > subsum) {
        right --
      } else {
        left ++
      }
    }
  }
  return result
};

console.log(threeSum([-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4]))
// console.log(threeSum([-1, 0, 1, 2, -1, -4]))
// console.log(threeSum([]))
// console.log(threeSum([0]))

File: 000001-000200/000015.三数之和/15.go
package main

import (
	"fmt"
	"sort"
)

// 方法一：用时 28 ms
// 执行用时超过了 93 %
// 消耗内存超过了 77 %
// 难度: 中等
func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	ret := [][]int{}
	for i := 0; i < len(nums) - 2; i++ {
		if nums[i] > 0 {
			break
		}
		if i > 0 && nums[i] == nums[i - 1] {
			continue
		}
		for left, right := i + 1, len(nums) - 1; left < right; {
			if nums[left] + nums[right] == -nums[i] {
				ret = append(ret, []int{nums[i], nums[left], nums[right]})
				left ++
				right --
				for ; nums[left] == nums[left - 1] && left < right; {
					left ++
				}
				for ; nums[right] == nums[right + 1] && left < right; {
					right --
				}
			} else if nums[left] + nums[right] > -nums[i] {
				right --
			} else {
				left ++
			}
		}
	}
	return ret
}

func main ()  {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
	fmt.Println(threeSum([]int{}))
	fmt.Println(threeSum([]int{0}))
	fmt.Println(threeSum([]int{0, 0, 0}))
}
